File: modules/padel-api/utils/granularity.go
package utils

import "time"

const DailyGranularityMaxDays = 45

func DetermineGranularity(from, to time.Time) (string, int) {
	daysDiff := int(to.Sub(from).Hours()/24) + 1
	if daysDiff <= 0 {
		daysDiff = 1
	}
	if daysDiff == 1 {
		return "hourly", daysDiff
	}
	if daysDiff <= DailyGranularityMaxDays {
		return "daily", daysDiff
	}
	return "monthly", daysDiff
}

func GranularityLabelFormat(granularity string) string {
	switch granularity {
	case "hourly":
		return "HH24:00"
	case "daily":
		return "DD Mon"
	default:
		return "Mon YYYY"
	}
}

func GranularityTrunc(granularity string) string {
	switch granularity {
	case "hourly":
		return "hour"
	case "daily":
		return "day"
	default:
		return "month"
	}
}

File: modules/web-padel/src/utils/granularity.ts
export const DAILY_GRANULARITY_MAX_DAYS = 45;

export type Granularity = 'hourly' | 'daily' | 'monthly';

export function determineGranularity(from: Date, to: Date): { granularity: Granularity; daysDiff: number } {
  const msPerDay = 24 * 60 * 60 * 1000;
  const daysDiff = Math.round((to.getTime() - from.getTime()) / msPerDay) + 1;
  if (daysDiff <= 0) {
    return { granularity: 'hourly', daysDiff: 1 };
  }
  if (daysDiff === 1) {
    return { granularity: 'hourly', daysDiff };
  }
  if (daysDiff <= DAILY_GRANULARITY_MAX_DAYS) {
    return { granularity: 'daily', daysDiff };
  }
  return { granularity: 'monthly', daysDiff };
}

export function granularityLabelFormat(granularity: Granularity): string {
  switch (granularity) {
    case 'hourly':
      return 'HH:00';
    case 'daily':
      return 'DD MMM';
    case 'monthly':
      return 'MMM YYYY';
  }
}

export function granularityTrunc(granularity: Granularity): string {
  switch (granularity) {
    case 'hourly':
      return 'hour';
    case 'daily':
      return 'day';
    case 'monthly':
      return 'month';
  }
}
